Plot Range-Time

mts299 opened the pull request:
The initial start for generating summary plots; this PR focuses on range-time parameter plots which plot array parameters along a date x-axis in color mesh representation. The README is updated to give a couple of examples, and test/unit/rtp.py (dependent on the file name) has several various plotting examples in it if you would like to look at those. 

```
import pydarn
import matplotlib.pyplot as plt 

fitacf_file = "./20181209.C0.sas.fitacf"
darn_read = pydarn.DarnRead(fitacf_file)
fitacf_data = darn_read.read_fitacf()

pydarn.RTP.plot_range_time(fitacf_data, parameter='elevation',
                           beam_num=7)
plt.show()
```

![pydarn_simple_plot](https://user-images.githubusercontent.com/6520530/56516973-0977ef00-64f9-11e9-9964-785b28ff87ae.png)

This plotting method is an extension of `matplotlib.pyplot` so you can still manipulate the overall plot outside of the method call. Thus this method will be a bare-bones representation of range-time plot. 

It also can take in matplotlib axes objects:

```
import pydarn
import matplotlib.pyplot as plt 

fitacf_file = "./20181209.C0.sas.fitacf"
darn_read = pydarn.DarnRead(fitacf_file)
fitacf_data = darn_read.read_fitacf()

fig, (ax1, ax2) = plt.subplots(2, 1)
pydarn.RTP.plot_range_time(fitacf_data, parameter='elevation',
                           beam_num=7, ax=ax1, channel=1,
                           date_fmt='')
ax1.set_ylabel("ch. 1")

pydarn.RTP.plot_range_time(fitacf_data, parameter='elevation',
                           beam_num=7, ax=ax2,
                           channel=2)
ax2.set_xlabel("Date (UTC)")
ax2.set_ylabel("ch. 2")

fig.suptitle("RTP subplots using Axes Object")
plt.show()
```
![pydar_axes_range_time_plot](https://user-images.githubusercontent.com/6520530/56517230-a470c900-64f9-11e9-9385-202aaf2d53c5.png)

Further documentation on this method and others will be done in another branch as this is not being pushed in `develop` right now. I will need testers to ensure the plots are correct and reviewers of the code.

aburrell commented:
Will we be able to select a subset of data to plot based on other fit parameters?  As an easy one, let's say I only want to plot velocities with elevation angles below 20 degrees.  Or are we stuck plotting everything from the fitacf data object?

mts299 commented:
@aburrell I just committed some code that handles this case. 
Here is pruning example:
```python
import pydarn
import matplotlib.pyplot as plt 

fitacf_file = "./20181209.C0.sas.fitacf"
darn_read = pydarn.DarnRead(fitacf_file)
fitacf_data = darn_read.read_fitacf()
data = pydarn.dmap2dict(fitacf_data)

fig, (ax1, ax2) = plt.subplots(2, 1)
pydarn.RTP.plot_range_time(fitacf_data, parameter='velocity',
                           beam_num=7, ax=ax1, color_map='jet_r',
                           boundary=(-200, 200),
                           date_fmt='')
ax1.set_ylabel("range gates")

pydarn.RTP.plot_range_time(fitacf_data, parameter='velocity',
                           beam_num=7, ax=ax2, color_map='jet_r',
                           boundary=(-200, 200), max_array_filter={'elv': 20})
ax2.set_ylabel("range gates")

fig.suptitle("Whole fitacf Velocity data vs. Elevation <= 20 degrees")
plt.show()
```

![pydarn_pruning_example](https://user-images.githubusercontent.com/6520530/56536144-91babc00-651a-11e9-8df1-881c16e77acf.png)

I have also included minimum, maximum for both array values and scalars. You can also have multiple parameters to filter over. 

Please note: I have only implemented Range Gates so far just to keep things simple.

aburrell commented:
Not a bad start.  Something odd happened when I changed the x-range to be bigger and smaller (the top figure didn't recover all of its ticks), but other than that the plotting behaves sensibly. 

 I am not a fan of not passing back the figure and axis handles when you don't declare them before plotting, but that's easy enough to get around so I don't see any reason to change your output.  It also makes it clear that you should be declaring the figure and axis if you want to mess around with things, so that sensibly limits the scope of the routine.

Initially I was confused as to why I wasn't getting the default colorbar labels.  Then I realized it was because you require the long names of the parameters for that functionality rather than the short names.

I ran a test case based off of your examples and got:

```
import pydarn
import matplotlib.pyplot as plt
import datetime as dt

fitacf_file = '20170120.0001.00.inv.fitacf'
stime = dt.datetime(2017,1,20)
etime = dt.datetime(2017,1,20,2)
sd_read = pydarn.DarnRead(fitacf_file)
fitacf_data = sd_read.read_fitacf()

f = plt.figure()
ax = f.add_subplot(211)
ax1 = f.add_subplot(212)

pydarn.RTP.plot_range_time(fitacf_data, parameter='v', beam_num=7, ax=ax, color_map='Spectral_r', boundary=(-150, 150), date_fmt='%H:%M', max_array_filter={'gflg':1})
ax.set_facecolor('0.9')
ax.set_ylabel("Range Gate")
ax.set_xlim(stime, etime)

out = pydarn.RTP.plot_range_time(fitacf_data, parameter='elv', beam_num=7, ax=ax1, color_map='Spectral_r', boundary=(0, 50), date_fmt='%H:%M', equal_array_filter={'gflg':1}, color_bar_label='$\Delta (^\circ)$')
ax1.set_xlabel("Universal Time (HH:MM)")
ax1.set_ylabel("Range Gate")
ax1.set_xlim(stime, etime)
ax1.set_facecolor('0.9')
f.suptitle("INV Test for {:}, Beam 7".format(stime.date()))
f.subplots_adjust(right=.99)
```

![pydarn_rti_inv](https://user-images.githubusercontent.com/7045886/56588753-fbd27000-65b1-11e9-88e1-c43b71cd9e59.png)

aburrell commented:
![pydarn_rti_inv](https://user-images.githubusercontent.com/7045886/56588975-5d92da00-65b2-11e9-87a3-6bfd2ef72c9f.png)

mts299 commented:
@aburrell Thank you for getting to testing it so quickly :) 

If you let me know what you tried to do for resizing the x-range, I can see if there is anything I can do in the code to change that behaviour. It may have to do with me setting the x-axis in the mthod to get the date format. Depending if we want that in the method or not, may allow us to have more freedom with the x-axis outside of the method. 

In regards to not returning the `figure` and `axes` objects, I did this as a desigm choice to try and truel make the plotting methods an extension of matplotlib plotting. My method does return `matplotlib.collections.QuadMesh`, which is what you would get from calling pcolormesh, the color bar object, and the color map object. However, for the full summary plot implementation I will probably return the `figure` and `axes` object. I will also make note of this in the documentation. 

Yes, I did the long name in relation to DaViTpy doing it like that. However, if this is more of annoyance then useful feature we could just remove it?

aburrell commented:
@mts299 I tried resetting xlim several times, once too large, once too small, and then back to the original limits.  It wasn't able to recover the labels after the too small (I don't recall what the labels looked like when it was too large).  This is probably worth fixing, since the ability to zoom and then expand the time range is pretty commonly desired.

As for the long name - I find it annoying, because I have to remember two names: what the routine wants as input and what the data variables are.  It's easier for me to look up the data keys in the fitacf object and plot what that shows me.  It's probably worth asking others what they think.

mts299 commented:
@aburrell I tried changing the x-axis range with `xlim` and `set_xlim` and the labels looked okay for me:
![x-axis_change](https://user-images.githubusercontent.com/6520530/56843567-986d6a00-685f-11e9-8561-3b436b3b3d98.png)

here is my code:
```python
import pydarn
import matplotlib.pyplot as plt
from datetime import datetime

darn_read = pydarn.DarnRead("20140101.0401.00.han.fitacf")
data = darn_read.read_fitacf()



plt.title("change of axis")
plt.subplot(3, 1, 1,)
pydarn.RTP.plot_range_time(data, beam_num=7)

plt.subplot(3, 1, 2)
pydarn.RTP.plot_range_time(data, beam_num=7)
plt.xlim(datetime(2014, 1, 1, 5), datetime(2014, 1, 1, 6))

plt.subplot(3, 1, 3)
pydarn.RTP.plot_range_time(data, beam_num=7)
plt.gca().set_xlim(datetime(2014, 1, 1, 0), datetime(2014, 1, 1, 8))

plt.show()
```

Can you let me know if this is what you got and what the problem is?

aburrell commented:
I replicated a plot similar to yours and tried to break it and couldn't.  Let's chalk up the error to gremlins and leave it me.

aburrell commented:
No, I did it!

Here's what you have to do:

```
import pydarn
import matplotlib.pyplot as plt
from datetime import datetime

darn_read = pydarn.DarnRead("20170120.0001.00.inv.fitacf")
data = darn_read.read_fitacf()

f = plt.figure()
ax = f.add_subplot(111)
pydarn.RTP.plot_range_time(data, beam_num=7, ax=ax)

# To replicate the error, you need to change the date format
ax.xaxis.set_major_formatter(mpl.dates.DateFormatter('%H:%M'))

# Then you have to extend the range to be larger than the time period
ax.set_xlim(datetime(2017, 1, 20, 1), datetime(2017, 1, 25, 2))

# And then go back and fix it
ax.set_xlim(datetime(2017, 1, 20, 1), datetime(2017, 1, 20, 2))
```

mts299 commented:
@aburrell I think we are hitting that cell phone paradox :p 
Besides the errors I got from your script not declaring what `mpl` was I got it to work and produce a plot that I think is correct:

```python
import pydarn
import matplotlib.pyplot as plt
from matplotlib import dates
from datetime import datetime

darn_read = pydarn.DarnRead("20180220.0001.00.rkn.3.0.fitacf")
data = darn_read.read_fitacf()

f = plt.figure()
ax = f.add_subplot(111)
pydarn.RTP.plot_range_time(data, beam_num=7, ax=ax)

# To replicate the error, you need to change the date format
ax.xaxis.set_major_formatter(dates.DateFormatter('%H:%M'))

# Then you have to extend the range to be larger than the time period
ax.set_xlim(datetime(2018, 2, 20, 1), datetime(2018, 2, 25, 2))

# And then go back and fix it
ax.set_xlim(datetime(2018, 2, 20, 1), datetime(2018, 2, 20, 2))

plt.show()
```

![replicate_error](https://user-images.githubusercontent.com/6520530/57388564-b0a18b00-7175-11e9-99ed-0b4a4e2ed60f.png)

I added a third plot to demonstrate how you can use the `date_fmt` key in the plotting method to change the x-axis format.

aburrell commented:
Ah, I forgot I had also imported mpl.  Sorry!

Huh, how frustrating that it works for you but not for me.  Regardless, unless more people start complaining I think this can be put aside as a curiosity.  It is easy for me to get around, I can just make sure I change the date range before adjusting the format.

mts299 commented:
At CEDAR you can show me and I can see what is going on. I am curious as well :) 

In terms of those parameter names, does anyone agree we should get rid of them?

asreimer commented:
Planning on looking at this over the weekend!

mts299 commented:
Hey, I removed the parameter names by @aburrell suggestion (with no argument against it). 
I have also added normalization dependency injection for plotting different color normalizations. 

Here is an example using Borealis data: 
```python
import pydarn
import matplotlib.pyplot as plt 
from matplotlib import colors

rawacf_file = "20190516.0200.01.sas.0.rawacf.rawacf.dmap"
darn_read = pydarn.DarnRead(rawacf_file)
rawacf_data = darn_read.read_rawacf()

lognorm = colors.LogNorm

pydarn.RTP.plot_range_time(rawacf_data, beam_num=0, parameter='pwr0',
                           color_bar_label="pwr0", color_norm=lognorm)

plt.title("Borealis RAWACF pwr0 plot for 2019 05 23")
plt.ylabel('Range Gates')
plt.xlabel('Date (UTC)')
plt.show()
```

![borealis_pwr_plot](https://user-images.githubusercontent.com/6520530/58493006-499d4380-812f-11e9-9715-0700dd5321e3.png)

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+
+"""
+Range-time Intensity plots
+"""
```
I recommend saying "Range-Time Parameter (a.k.a Intensity) plots"

mts299 replied:
Done

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+
+class RTP():
+    """
+    Range-time intensity plots SuperDARN data using the following fields:
```
Change intensity to parameter here and elsewhere in the code

mts299 replied:
Done

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+import numpy as np
+from typing import List
+from datetime import datetime, timedelta
+
```
PEP8 the import order here and throughout (as before)

mts299 replied:
done

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+
+    @classmethod
+    def plot_range_time(cls, dmap_data: List[dict], *args,
+                        parameter: str = 'p_l', beam_num: int = 0, ax=None,
```
Stylistically, I recommend removing the whitespace about the '=' here

mts299 replied:
Flake 8 PEP 8 linter disagrees. I am just going to follow Flake 8 because I encourage others to use it to find syntax errors and PEP 8 violations.

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+                        parameter: str = 'p_l', beam_num: int = 0, ax=None,
+                        color_norm = None, **kwargs):
+        """
+
```
Missing description of the function

mts299 replied:
done

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+                    'date_fmt': "%y/%m/%d\n%H:%M",
+                    'color_bar': True,
+                    'color_bar_label': '',
+                    'color_map': 'jet',
```
Let's be bold and change the default colormap.  If we want to keep the red/blue we can use Spectral_r or seismic_r

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+        y_max = cls.dmap_data[0]['nrang']
+
+        # z: parameter data mapped into the color mesh
+        z = np.zeros((1, y_max)) * np.nan
```
I think it's clearer to use `np.full((1, y_max), fill_value=np.nan)` here and elsewhere

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+            if diff_time > 2.0:
+                # if there is gap data (no data recorded past 2 minutes)
+                # then fill it in with white space
+                for _ in range(0, int(np.floor(diff_time/2.0))):
```
Do you need to start range with zero?  Or is this stylistic?

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+
+            # Get data for the provided beam number
+            if (beam_num == 'all' or dmap_record['bmnum'] == beam_num) and\
+               (settings['channel'] == 'all' or
```
I believe you don't need the `\` for a line break within a parenthetical expression.

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+                            if settings['ground_scatter'] and\
+                               dmap_record['gflg'][j] == 1:
+                                # chosen value from davitpy to make the
+                                # ground scatter a different color
```
groundscatter is one word

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+
+    @classmethod
+    # TODO: could parallelize this method
+    def __filter_data_check(cls, dmap_record: dict, settings: dict, j: int) -> bool:
```
Recommend changing `j` to `islist` or `irg` to be more descriptive.

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+        -------
+        pass_flg : bool
+            boolean indicating if the dmap_record passes all filter checks
+        """
```
Recommend adding some comments here above each for-loop that says what it is doing.

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+
+        Parameters
+        ----------
+        paraneter: str
```
Misspelled

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+
+        Returns
+        -------
+        datetime:
```
Since you are returning a function call perhaps say `datetime object` and omit the colon?

mts299 replied:
done

aburrell commented on pydarn/plotting/rtp.py:
```diff
@@ -0,0 +1,438 @@
+                        minute=minute, second=second, microsecond=micro_sec)
+
+    # Needs its own method because it generates vertical lines when
+    # the cpid changes
```
Are these all `TODO`?

mts299 replied:
HAHA! this is older code. Deleted :p 

However, if we want to have different visuals for CP IDs it would make sense to remove it from time-series. Something to think about!

aburrell commented on test/unit/test_rtp.py:
```diff
@@ -0,0 +1,247 @@
+        """
+        pydarn.RTP.plot_range_time(self.fitacf_data, parameter='p_l',
+                                   beam_num=7, ground_scatter=True)
+        plt.title("Elevation with Ground Scatter")
```
Change Ground Scatter to either Ground Backscatter or Groundscatter

mts299 replied:
Done

aburrell commented on test/unit/test_rtp.py:
```diff
@@ -0,0 +1,247 @@
+        ax.set_xlabel("Date (UTC)")
+        ax.set_ylabel("Elevation $degrees$")
+        fig.suptitle("RTP plot using Axes object")
+        plt.show()
```
Do we want to show things during tests?  I've run on some computers where I could only write out image files, not display them.

mts299 replied:
Typically not, almost defeats the purpose of `python test`. However, the unit tests for plotting becomes very simplistic and testing becomes more visual.

Right if you `ssh` or don't have DISPLAY enabled on the machine then show() will not work. This was really to help me test it and ensure all the plots looked good. 

I think I will be making a discussion revolving around `python test` and I will include the plotting tests and what people would prefer as developers.

aburrell commented on test/unit/test_rtp.py:
```diff
@@ -0,0 +1,247 @@
+        fitacf_file = "../testfiles/20180220.C0.rkn.fitacf"
+        darn_read = pydarn.DarnRead(fitacf_file)
+        cls.fitacf_data = darn_read.read_fitacf()
+
```
Do you need a tearDown function?

mts299 replied:
tearDown is good for disconnecting connections to databases or closing files if you open them. However; I do not necessarily need one because we don't have any open file sockets in this case. I could have one to delete the data but python's garbage collector should do that. I was thinking if we saved images we could make some tearDown type function such that the images don't clutter up the folder.
